src/uct/cuda/cuda_ipc/cuda_ipc_cache.c

```diff
@@ -205,36 +239,79 @@ uct_cuda_ipc_open_memhandle_mempool(uct_cuda_ipc_rkey_t *key,
                                     CUdeviceptr *mapped_addr)
 {
     CUmemAccessDesc access_desc = {};
-    CUdeviceptr dptr;
-    ucs_status_t status;
+    ucs_status_t status         = UCS_OK;
+    CUmemoryPool *mpool;
+    khiter_t khiter;
+    int khret;
 
-    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportFromShareableHandle(
-                &key->ph.pool, (void *)&key->ph.handle.fabric_handle,
-                CU_MEM_HANDLE_TYPE_FABRIC, 0));
-    if (status != UCS_OK) {
-        return status;
-    }
+    pthread_rwlock_rdlock(&uct_cuda_ipc_rem_mpool_cache.lock);
+    khiter = kh_get(cuda_ipc_rem_mpool_cache,
+                    &uct_cuda_ipc_rem_mpool_cache.hash,
+                    (char*)&key->ph.handle.fabric_handle);
+    if (ucs_likely(khiter != kh_end(&uct_cuda_ipc_rem_mpool_cache.hash))) {
+        key->ph.pool = kh_val(&uct_cuda_ipc_rem_mpool_cache.hash, khiter);
 
-    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportPointer(&dptr,
-                key->ph.pool, (CUmemPoolPtrExportData*)&key->ph.ptr));
-    if (status != UCS_OK) {
-        return status;
+        status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportPointer(mapped_addr,
+                    *key->ph.pool, (CUmemPoolPtrExportData*)&key->ph.ptr));
+        goto err;
     }
 
-    status = uct_cuda_ipc_init_access_desc(&access_desc);
-    if (status != UCS_OK) {
-        return status;
+    /* acquire writer lock as we did not find cached mpool in hashmap */
+    pthread_rwlock_unlock(&uct_cuda_ipc_rem_mpool_cache.lock);
+    pthread_rwlock_wrlock(&uct_cuda_ipc_rem_mpool_cache.lock);
```

Comment:
Just an idea, I'm not requesting this change: 
This is correct logic, but can we do better in terms of scalability?
What if we don't acquire a write lock until we construct a new object?
This way we drastically reduce the scope of the write lock - it will be held only during khash update basically:
```
pthread_rwlock_unlock
// Construct new object
mpool = ucs_malloc(sizeof(*mpool), "cuda_ipc_mempool");
cuMemPoolImportFromShareableHandle
uct_cuda_ipc_init_access_desc
cuMemPoolSetAccess
// now acquire the write lock and update the hashmap
pthread_rwlock_wrlock
khiter = kh_put
// Here we need to check if khash is updated already for this key
if (khret == UCS_KH_PUT_KEY_PRESENT)
  ucs_free(mpool); // Release tmp mpool, and use a value from hash
```
The pros is scalability, the cons is more complexity and potentially more CPU cycles (can lead to parallel object creation for the same key). Just to consider, up to you

Reply:
Thanks for the suggestions @iyastreb. 

Multiple threads calling `ImportFromShareableHandle` is undesirable because: 
1. This PR is aiming to avoid multiple calls to ImportFromShareable/SetMemAccess as they are expensive by caching the calls to these operations. Not having a lock/unlock around these calls would negate that purpose.
2. It would lead to multiple mappings of the same remote allocation and all but one thread have to release their mappings after the subsequent lookup. Ultimately only one of the threads has done useful work.

As the costly critical path is stressed once per unique key, IMO it's better to keep existing logic. Let me know if you think there is still room for optimization.